### DISCUSSION TITLE
Exported enums should remember set key, not values.

### DISCUSSION THREAD

**Discussion by NickSw1ft** [godotengine/godot-proposals#9128] [Engine Core]
If you change order of enum keys in script, all exported variables of this enum are going to be shifted.
E.g. in the following example if I set `boo` to `C` via editor and then change order or insert element in `Foo`, `boo` key will change.
`enum Foo {A, B, C}`
`@export var boo: Foo = Foo.A`
I would like to propose that in such cases exported variable is assigned a value corresponding to key set in editor.

**Comment by dalexeev** [comment]
This is a core issue, built-in enums also store values, not keys (so we try not to shift values). This could be solved with a new usage flag and annotation, but if I remember correctly, the built-in enums do not store keys in release builds to save space.

As a workaround, you can explicitly specify the enum values:

```gdscript
enum Foo {
    A = 0,
    B = 1,
    C = 2,
    # Add new values to the end, don't shift IDs!
}
@export var boo: Foo = Foo.A
```

or use the `@export_enum` annotation:

```gdscript
@export_enum("A", "B", "C") var boo: String = "A"
```

**Reply by NickSw1ft** [reply]
Thank you for suggested workarounds. I would like to point out that secong one is really awkward as it takes away the convenience of enums and editor hints. The first one is neat, but could be really confusing and clunky after a dozen of values and a couple of shiftings.

**Comment by Mickeon** [comment]
If you remember the keys instead of the values, the problem is not solved, but it is shifted to the keys. If you change the name of any key the same issue happens.
Storing the raw value is more efficient and less of a workaround (since enum values are just integers at the end of the day).

The editor should also not be trusted in predicting what has been changed and how. This is the kind of thing that some sort of refactoring tool could help manage (something that has been asked many times before).

**Reply by Ohan17** [reply]
Saving enum keys instead of the values does solve some aspect of current enum:

1. It will not be shifted when reordering enums
2. There is a chance to add warnings to catch invalid enum at load time which is still better than it silently reference a different key or break entirely at runtime.

Both of which makes it extreamly safer to use enum, also enum keys are StringName which I don't think there is ever a situtation where that will be a performance issue compared to using int.

**Comment by lupuchard** [comment]
I don't like the current behavior because it makes scene files inconsistent with GDScript (or any other language with enums). In code, enums are referenced via their names, only becoming integers at runtime, this is the primary benefit of enums which is why they can normally be reordered freely and the programmer is generally not supposed to be concerned with the exact integer each enum value will compile to (that's why those values are implicit by default, unlike with constants). By referencing enums as integers, scene files break this convention which is IMO unexpected, hurts version-control readability, and defeats a lot of the benefits of enums.

The argument I'm not understanding is that since enums are integers at runtime, they have to be integers in the resource files? But every type has their own means of serialization. A Vector3 isn't the string "Vector3(16.9206, 1, 20.5)" at runtime so why should enums have to be serialized as integers?

To handle the issue of renaming enum values, scenes could potentially store int/string pairs and fall back to 'same index' if none of the names match. Or utilize whatever solution is found to the existing problem of renaming exported values (https://github.com/godotengine/godot-proposals/issues/3152) since this would essentially be the same thing.
